### ISSUE TITLE
Create fields for structural block types

### ISSUE THREAD

**Issue by alexgleason** [torchbox/wagtail#2048]
I want to make a field that's a simple list downloads. The following code accomplishes what I want using StreamField, but the API is a little weird.

```python
class ResourcesPage(Page):
    files = StreamField([
        ('files', blocks.ListBlock(DocumentChooserBlock()))
    ])
```
```
{% for file in self.files.value.files %}
  {{ file }}
{% endfor %}
```

Ideally it would look something like this:

```python
class ResourcesPage(Page):
    files = ListField(DocumentChooserBlock())
```
```
{% for file in self.files %}
  {{ file }}
{% endfor %}
```

Here I've created a new imaginary type, ListField. There are a few possible solutions (such changing the way StreamField parameters work) but I think it makes sense to have a wrapper field around each of the [structural block types](http://docs.wagtail.io/en/v1.2/topics/streamfield.html#structural-block-types).

Here's an example of how an implementation could look:

```python
from wagtail.wagtailcore.fields import StructField, ListField, StreamField
from wagtail.wagtaildocs.blocks import DocumentChooserBlock
from wagtail.wagtailimages.blocks import ImageChooserBlock
from wagtail.wagtailcore import blocks


class RecipePage(Page):
    # StructBlock wrapper
    chef = StructField([
        ('first_name', blocks.CharBlock(required=True)),
        ('surname', blocks.CharBlock(required=True)),
        ('photo', ImageChooserBlock()),
        ('biography', blocks.RichTextBlock())
    ])
    
    # ListBlock wrapper
    ingredients_list = ListField(blocks.CharBlock(label="Ingredient"))
    
    # StreamBlock (just a regular StreamField)
    instructions = StreamField([
        ('paragraph', blocks.RichTextBlock(icon="pilcrow")),
        ('h1', blocks.CharBlock(classname="title", icon="title"))
    ])
```

**Comment by timheap** [comment]
StreamFields are great, as they get around a limitation of databases, where storing a collection of different things is difficult. This flexibility is not required for most of your example, which is better served by creating a model normally, possibly using the new [Django PostgreSQL `ArrayField`](https://docs.djangoproject.com/en/1.9/ref/contrib/postgres/fields/#arrayfield):

```python
class RecipePage(Page):
    chef_first_name = models.CharField()
    chef_surname = models.CharField()
    chef_photo = models.CharField()
    chef_biography = models.ForiegnKey('wagtailimages.Image', null=True, on_delete=models.SET_NULL)

    ingredients_list = ArrayField(models.CharField())

    instructions = StreamField([ ... ])
```

This has the added benefit of guarantees around database integrity, with the `chef_photo` field being set to null when the photo is deleted. `StreamField`s do not offer this, as they are just a blob of text as far as the database is concerned.

An `ArrayField` would not handle a more complicated ingredient definition though, which is where something that could handle `ListField(blocks.StructBlock([ ... ]))` would be useful. Currently this would require a separate model and an `InlinePanel`, or a `StreamField` with a single type of block.

```python
class RecipePage(Page):
    # ...
    ingredients_list = ListField(blocks.StructBlock([
        ('ingredient', blocks.CharField()),
        ('amount', blocks.CharField()),
    ]))

**Comment by alexgleason** [comment]
Bringing this back up, because I do need a ListField for the project I'm working on.

1. My page is divided into sections
2. Users must be able to add, rearrange, and delete sections
3. Sections contain StreamField content

In this case, it makes sense to have something like:

```python
body = ListField(blocks.StructBlock([
    ('title', blocks.CharBlock()),
    ('content', MyStreamBlock())
]))
```

**Comment by gasman** [comment]
The core team has discussed this today, and we're happy with the principle of adding `ListField` / `StructField` types in parallel with StreamField (or, perhaps more simply, allowing StreamField to accept `ListBlock` / `StructBlock` as the top-level block).

**Comment by Adrian-Turjak** [comment]
+1

This would be so very very useful. I'm willing to be this will get a lot of use of it existed.

I already have a streamfield block which already does what I want, but I need a given set of pages to have a single optional instance of said block. Trying to achieve the same thing using Django models and an inline field just isn't possible, or is very painful.

A BlockField() would solve my problem for me. The GUI would make sense, and it would allow reusing the existing block template.

If someone points me where to start I'd gladly help with this, or am more than happy to help review or debug any Work towards this.

**Comment by MShel** [comment]
I'll work on that.

**Comment by stuaxo** [comment]
Do any of these things exist yet ?

**Comment by kutenai** [comment]
I just asked about this on Slack and was pointed here.
This would be a nice to have feature that would make the models a little cleaner. My use case is that I have a collection of pre-defined block types that go into a technical content page. I want to be able to add blocks as fixed blocks (hence the need for a StructBlock), or stream fields, which already works beautifully.

The options available are:
-  StreamField -- which allows N blocks, when I want just 1, 
- Individual Fields for each "block", grouped using a MultiFieldPannel -- this isn't a re-usable block
- External Model, with ForeignKey.. not even sure if there is a 1-1 key here?

A "StructField" could take a block, similar to a StreamField, but accept just a single block, hence it would work just like a stream field from a DB perspective, but have a single block item, rather than an array of items as a StreamField does. 

Just thinking out loud

**Comment by lacymorrow** [comment]
I believe I am working on a use-case that would benefit greatly from what has been described above. 

I have a legal page with content and asides in a two-column layout. I need to visually represent this two-column layout in the admin UI, so that a user may add blocks which contains exactly one aside and one content section.

![image](https://user-images.githubusercontent.com/1311301/39442171-3cee4f4e-4c65-11e8-8ee5-052148fde6ca.png)

**Comment by ddiazpinto** [comment]
In my opinion, the best approach is to have an specific field for each **structural block type**:
 - ```StructField``` for ```StructBlock```
 - ```ListField``` for ```ListBlock```
 - ```StreamField``` for ```StreamBlock```

Regarding to the wagtail admin views:
 - ```StructField``` wouldn't need to include a "new" button, just render the block form.
 - ```ListField``` wouldn't need to show the *stream menu* to choose the model to add.
 - ```StreamField``` would work as usual

Of course ```ListField``` could be considered just a particular case of ```StreamField``` and be omitted, but if ```ListBlock``` exists, I think ```ListField``` should exists too.

**Comment by Benoss** [comment]
This would be an awesome feature to have for the upcoming LTS :)

**Comment by StriveForBest** [comment]
Any updates on this?

**Comment by mabdullahabid** [comment]
Updates on this?

**Comment by DIGIREN** [comment]
Also wondering if there are any updates on this? It would be a useful ability

**Comment by pdcribeiro** [comment]
Also hoping for this!

> In my opinion, the best approach is to have an specific field for each **structural block type**:
> 
>     * `StructField` for `StructBlock`
> 
>     * `ListField` for `ListBlock`
> 
>     * `StreamField` for `StreamBlock`

**Comment by Fcog** [comment]
I was hoping this feature was out already while scrolling down the page. :(

**Comment by richadr** [comment]
For a site we are developing now, I created a "workaround" `StructField` class that can be used to add `StructBlock`s directly to your page.  It feels a bit hacky, but makes my page models so much cleaner.  Note: not tested in production, and we have a headless setup so I'm only interested in the API output (I did not check how this behaves in a template).

```python
from wagtail.admin.edit_handlers import StreamFieldPanel
from wagtail.core import blocks
from wagtail.core.fields import StreamField


class StructField(StreamField):
    """
    A `StructField` allows you to place a single `StructBlock` as a field in your page model, where
    normally `StructBlock`s can only be contained inside a `StreamField`.

    Use like you would add any other field, e.g. `example = StructField(ExampleBlock, blank=True)`.

    Note: this is actually just a `StreamField` under the hood (with a `max_num` of 1 and a
    modified `get_api_representation` to output the value directly instead of an array).  Your
    `StructBlock`s should have a label, otherwise the page editor will display "_structvalue".
    """

    def __init__(self, structblock, blank=False, **kwargs):
        def get_api_representation(self, value, context=None):
            if value is None or len(value) is 0:
                return None

            return value[0].block.get_api_representation(
                value[0].value, context=context
            )

        # Dynamically create a class based on StreamBlock, that contains just one field called
        # `_structvalue` (this name is unused in the StructField API output; the value itself is
        # outputted directly).
        StructValueStreamBlock = type(
            "StructValueStreamBlock",
            (blocks.StreamBlock,),
            {
                "_structvalue": structblock,
                "get_api_representation": get_api_representation,
            },
        )

        super().__init__(
            StructValueStreamBlock(max_num=1, required=not blank), blank=blank, **kwargs
        )


class StructFieldPanel(StreamFieldPanel):
    pass
```

**Comment by stuaxo** [comment]
If the code is shared between the new field types and the original structural blocks, maybe moving the current code to a base type would make more sense, then have both the the Field and Block types extend it.

It might make sense to generate both, as above.

Otherwise, the object hierarchy becomes a bit weird.

**Comment by ssyberg** [comment]
Following

**Comment by Navaneeth-Sharma** [comment]
Any updates on this ?

**Comment by thibaudcolas** [comment]
Reviewed today – assigning to @gasman who will be able to provide an update on where we’re at with this (#9738 takes us closer).

**Comment by gasman** [comment]
Have done some testing today to see how close we are to supporting blocks other than `StreamBlock` as the top-level block.

Testing with this page definition on bakerydemo:

```python
class CaptionedImageBlockPage(Page):
    body = StreamField(CaptionedImageBlock())
    content_panels = Page.content_panels + [
        FieldPanel("body"),
    ]
```

the following patches are required to get as far as rendering the 'create' view:

In `StreamField.__init__`, avoid passing the field's `blank` argument on to the block's `required` argument if the block doesn't support that:

```diff
-        # for a top-level block, the 'blank' kwarg (defaulting to False) always overrides the
-        # block's own 'required' meta attribute, even if not passed explicitly; this ensures
-        # that the field and block have consistent definitions
-        self.block_opts["required"] = not kwargs.get("blank", False)
+        if isinstance(block_types, Block) and not isinstance(block_types, StreamBlock):
+            # top-level block is something other than StreamBlock - don't set required flag
+            pass
+        else:
+            # for a top-level block, the 'blank' kwarg (defaulting to False) always overrides the
+            # block's own 'required' meta attribute, even if not passed explicitly; this ensures
+            # that the field and block have consistent definitions
+            self.block_opts["required"] = not kwargs.get("blank", False)
```

In `StreamField.get_default`, only use the field-level default if it's something non-falsy (if no explicit default is specified on the field, it will be the empty string, which `StreamBlock.normalize` can handle but not all block types will):

```diff
     def get_default(self):
-        return self.stream_block.normalize(super().get_default())
+        field_default = super().get_default()
+        if field_default:
+            return self.stream_block.normalize(field_default)
+        else:
+            return self.stream_block.get_default()
```

However, this now fails when submitting the form, with:

```
Exception Type: TypeError at /admin/pages/add/base/captionedimageblockpage/60/
Exception Value: Field 'id' expected a number but got <Image: Muddy Waters>.
```

I think this is where we run into Django's forms framework being less rigorous about what gets passed to `to_python` than it really should be - if I remember rightly, it's supposed to be invoked to convert database-format values into native Python ones, but also gets called during `clean`, on a value that's already native Python. StreamBlock's `to_python` / `normalize` has been battle-hardened against this to handle any plausible type Django might throw at it, but I'm not sure if that's the way we want to go for all block classes.

**Comment by joeyjurjens** [comment]
I just made a package for this for anyone that's still interested:
https://github.com/joeyjurjens/wagtail-block-fields

Follows the API of how it's described in the issue, so it's not as comment above where StreamField also supports non StreamBlocks.

It approaches it the same way how StreamField & StreamBlock work together. Was actually quite straight forward to implement, but I do not know if there are any edge cases yet (I'm using it, and seems to work fine!).

**Comment by danmanix** [comment]
Thanks @joeyjurjens - Just what I was looking for. 

I'm sure the UI could be stripped back further these field types were core, but it's an excellent example of the suggestion.
